src/translator/service.cpp

```diff
@@ -42,29 +71,58 @@ Service::Service(Ptr<Options> options, const void * model_memory)
   }
 }
 
-void Service::enqueue() {
+void Service::async_translate() {
   Batch batch;
   while (batcher_ >> batch) {
     pcqueue_.ProduceSwap(batch);
   }
 }
+#else  // WITH_PTHREADS
+void Service::initialize_async_translators() {
+  ABORT("Cannot run in async mode without multithreading.");
+}
+
+void Service::async_translate() {
+  ABORT("Cannot run in async mode without multithreading.");
+}
+#endif // WITH_PTHREADS
+
+std::future<Response> Service::translate(std::string &&input) {
+  Segments segments;
+  SentenceRanges sourceRanges;
+  text_processor_.process(input, segments, sourceRanges);
+
+  std::promise<Response> responsePromise;
+  auto future = responsePromise.get_future();
+
+  Ptr<Request> request = New<Request>(
+      requestId_++, /* lineNumberBegin = */ 0, vocabs_, std::move(input),
+      std::move(segments), std::move(sourceRanges), std::move(responsePromise));
+
+  batcher_.addWholeRequest(request);
+  if (numWorkers_ == 0) {
+    blocking_translate();
+  } else {
+    async_translate();
```

Comment:
Why are we exhausting the batcher at the end of every request?  Wasn't most of the point of the async batching thing that we could batch on the fly and mix from different requests?  

The batcher is likely its own thread, taking in requests and spitting out batches.

Reply:
Because we don't have asynchronous queueing yet? In the initial implementation in concurrent queuing, batcher's queuing using `std::async` transfers into a different thread (this is not merged yet). There are no locks or guards in place making batcher's internal container thread safe in this source to launch `std::async`.

Is there a change suggested here?

Reply:
I think you mean for me to move `addWholeRequest` duplicated into functions in preparation for concurrent queueing?

Reply:
Fine, I'm unhappy with this state of things but on the principle of making short focused pull requests we'll deal with that later.